Add protection for Order queries using a status that is not a registered post status.

nerrad opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

For background, you can see this pull made in the WC Blocks repo: https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/2912

Currently it is possible for filtered order statuses (via the `woocommerce_register_shop_order_post_statuses`) to be implemented incorrectly and thus resulting in unexpected results when doing a order query that is filtered by a status. The caller may _expect_ the status to be registered as a custom post status, when in fact it might not be because code executing later on that hook clobbers the order statuses actually registered. In turn, this means when the query makes its way to `WP_Query`, the non registered `post_status` will be removed from the query leading to potentially unexpected results.

This is potentially destructive if (as the case with logic in the WC Blocks plugin) the results from the query are used to delete orders.

The proposed protection in this pull does the following:

- If the query has a value on `post_status` query argument, then verify all statuses in that argument are a registered WP post status.
- If any of those are _not_ registered, then an exception is thrown. Otherwise processing continues.
- `WC_Order_Data_Store_CPT::query` catches any thrown exception and returns an empty array along with feeding the exception to `wc_caught_exception` which should surface in any error logs configured for the site.

### Some things to consider while reviewing

- The exception is being thrown from `WC_Order_Data_Store_CPT::get_wp_query_args` which is a protected method. I don't see this being used by any other code in Woo Core, but because this could feasibly be used in extension extending this class, if they aren't catching potential exceptions, and there is flawed code on sites, the thrown exception could be potentially breaking. If this is a concern, then before merging you may want to consider an alternative way of aborting the query.
- There's multiple other places an incoming query could be mangled (filters etc). Long term, it might be good to consider a similar feature to WP core for it's queries where query related filters can be explicitly suppressed via query_args. Having something like that is not a solution for the post_status option, but would be potentially good additional protection for queries needing to protect against filtering happening in WC data stores when making a query.

### How to test the changes in this Pull Request:

The changes seem to be covered well by existing tests (including new phpunit tests added). However, this does have impact on any code doing an Order query with order statuses. So smoke testing of UI/UX related to that, and also scanning extensions that might be doing that would probably be prudent.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Added protection for Order queries filtering by status which might not be a registered WP post status at the time of the query.

nerrad committed:
Add protection for order statuses used in a query that are not registered as a post status

nerrad committed:
Add tests for new order status verification in order queries

nerrad commented:
Just noting I see the linting fails and will get them fixed.

nerrad committed:
fix linting issues

nerrad commented:
Alright, code standard fails are fixed. The only failing test in this pull seems to be happening everywhere else too, so this should be ready for a review.

claudiosanches commented on includes/data-stores/class-wc-order-data-store-cpt.php:
```diff
@@ -772,15 +775,33 @@ protected function get_wp_query_args( $query_vars ) {
 				$query_vars['post_status'] = wc_is_order_status( 'wc-' . $query_vars['post_status'] ) ? 'wc-' . $query_vars['post_status'] : $query_vars['post_status'];
 			}
+
+			// verify post_status is actually registered.
```
```suggestion
			// Verify post_status is actually registered.
```

claudiosanches commented on includes/data-stores/class-wc-order-data-store-cpt.php:
```diff
@@ -772,15 +775,33 @@ protected function get_wp_query_args( $query_vars ) {
+						),
+						$post_status
+					);
+					// bail.
```
```suggestion
					// Bail.
```

claudiosanches commented on includes/data-stores/class-wc-order-data-store-cpt.php:
```diff
@@ -772,15 +775,33 @@ protected function get_wp_query_args( $query_vars ) {
+					$message = sprintf(
+						/* translators: 1: Status name */
+						__(
+							'The provided order query contains an order status that is not registered as a custom post status (status provided is %1$s). It is possible there is something removing any expected registered custom post status before the query is made.',
```
```suggestion
							'The provided order query contains an order status that is not registered as a custom post status (status provided is %s). It is possible there is something removing any expected registered custom post status before the query is made.',
```

claudiosanches commented on includes/data-stores/class-wc-order-data-store-cpt.php:
```diff
@@ -772,15 +775,33 @@ protected function get_wp_query_args( $query_vars ) {
+				$post_status_exists = get_post_stati( array( 'name' => $post_status ) );
+				if ( empty( $post_status_exists ) ) {
+					$message = sprintf(
+						/* translators: 1: Status name */
```
```suggestion
						/* translators: %s: status name */
```

claudiosanches commented on tests/legacy/unit-tests/order/class-wc-tests-order-functions.php:
```diff
@@ -15,6 +15,34 @@
+	 * Setup before tests
+	 */
+	public function setUp() {
+		// set listening for exceptions.
```
```suggestion
		// Set listening for exceptions.
```

claudiosanches commented on tests/legacy/unit-tests/order/class-wc-tests-order-functions.php:
```diff
@@ -15,6 +15,34 @@
+	}
+
+	/**
+	 * Teardown after tests
```
```suggestion
	 * Teardown after tests.
```

claudiosanches commented on tests/legacy/unit-tests/order/class-wc-tests-order-functions.php:
```diff
@@ -1539,4 +1567,19 @@ public function test_wc_generate_order_key() {
+	 * post status
+	 */
+	public function test_wc_get_order_non_registered_status_param() {
+		// temporarily hide error logging we don't care about (and keeps from polluting stdout).
```
```suggestion
		// Temporarily hide error logging we don't care about (and keeps from polluting stdout).
```

claudiosanches commented on tests/legacy/unit-tests/order/class-wc-tests-order-functions.php:
```diff
@@ -1539,4 +1567,19 @@ public function test_wc_generate_order_key() {
+		$this->assertEmpty( $orders );
+		$this->assertContains( 'provided order query contains an order status that is not registered', $this->caught_exception->getMessage() );
+		// restore original logging destination.
+		ini_set( 'error_log', $this->original_logging_destination );
```
```suggestion
		ini_set( 'error_log', $original_logging_destination );
```

claudiosanches commented on tests/legacy/unit-tests/order/class-wc-tests-order-functions.php:
```diff
@@ -1539,4 +1567,19 @@ public function test_wc_generate_order_key() {
+		$orders = wc_get_orders( array( 'status' => 'i-do-not-exist' ) );
+		$this->assertEmpty( $orders );
+		$this->assertContains( 'provided order query contains an order status that is not registered', $this->caught_exception->getMessage() );
+		// restore original logging destination.
```
```suggestion
		// Restore original logging destination.
```

claudiosanches reviewed (changes requested):
Changes look good, but there's some coding standards and some small bugs to fix.

nerrad committed:
Apply suggestions from code review

Multiple code style changes and fixing incorrect usage of object instance (via $this) in tests.

Co-authored-by: Claudio Sanches <contato@claudiosanches.com>

nerrad commented:
Thanks for the review @claudiosanches! I've applied the fixes and rebased this branch on master. Should be good to go for another review.

rrennick commented on includes/data-stores/class-wc-order-data-store-cpt.php:
```diff
@@ -772,15 +775,33 @@ protected function get_wp_query_args( $query_vars ) {
+					$message = sprintf(
+						/* translators: %s: status name */
+						__(
+							'The provided order query contains an order status that is not registered as a custom post status (status provided is %s). It is possible there is something removing any expected registered custom post status before the query is made.',
```
@nerrad Did you intend `an`  instead of `any` in `removing any expected registered`?

nerrad replied:
🤔 hmm ya `an` probably makes more sense here?

rrennick reviewed (changes requested):
Thanks @nerrad. Looks good other than that text change.

nerrad committed:
Fix minor grammar issue with notice message.

nerrad commented:
Requested text change fixed and ready for another review.

peterfabian commented:
@claudiosanches @rrennick can you guys please recheck if you're happy with the changes? Thanks!

vedanshujain commented on includes/data-stores/class-wc-order-data-store-cpt.php:
```diff
@@ -772,15 +775,33 @@ protected function get_wp_query_args( $query_vars ) {
+					$message = sprintf(
+						/* translators: %s: status name */
+						__(
+							'The provided order query contains an order status that is not registered as a custom post status (status provided is %s). It is possible there is something removing an expected registered custom post status before the query is made.',
```
Perhaps second part of the error message is not helpful, because [unregistered post statuses are definately removed in WP query](https://github.com/WordPress/WordPress/blob/b66ff2f68daf3a6c83154f2dd841c9a5dedd4c4e/wp-includes/class-wp-query.php#L2465).

vedanshujain commented on includes/data-stores/class-wc-order-data-store-cpt.php:
```diff
@@ -772,15 +775,33 @@ protected function get_wp_query_args( $query_vars ) {
+			$post_stati = is_array( $query_vars['post_status'] ) ? $query_vars['post_status'] : array( $query_vars['post_status'] );
+			foreach ( $post_stati as $post_status ) {
+				$post_status_exists = get_post_stati( array( 'name' => $post_status ) );
+				if ( empty( $post_status_exists ) ) {
```
should we collect these statuses in an array and then in the error message show all of them together (in case there are more than one un-registered statuses)?

nerrad replied:
Ya that sounds like a good improvement.

vedanshujain commented on includes/data-stores/class-wc-order-data-store-cpt.php:
```diff
@@ -853,28 +874,32 @@ protected function get_wp_query_args( $query_vars ) {
+				);
+			}
+		} catch ( \Exception $exception ) {
+			wc_caught_exception( $exception, __METHOD__ );
```
this function is deprecated (or at least is part of the file where functions go to die). let's throw the error.

nerrad replied:
Oh hmm, then ya it shouldn't be used. I'll leave it up to you folks how you implement error handling here.

vedanshujain reviewed (changes requested):
> if they aren't catching potential exceptions, and there is flawed code on sites, the thrown exception could be potentially breaking.

I would suggest that we support another field in `query_vars` called `strict`. If set to true, only then these changes will come into effect, this will allow us to be more liberal with throwing exceptions instead of silently ignoring them and adding to log. wdyt?

nerrad commented on includes/data-stores/class-wc-order-data-store-cpt.php:
```diff
@@ -772,15 +775,33 @@ protected function get_wp_query_args( $query_vars ) {
+						$post_status
+					);
+					// Bail.
+					throw new \Exception( $message );
```
Agreed. Since that has wider impact/use - I think you folks should take the lead on that. Also, is that something that could be done in a follow-up?

nerrad commented:
Thanks for the feedback @vedanshujain. Realistically, I'm likely not going to be able to carve out time to address the feedback anytime soon as our team's plate is currently fairly full with other things. I understand you folks will have to prioritize this on your end as well.

I only ever intended to surface the problem and a potential solution for it. For now, we have some protection in the blocks code but it's an issue that could surface in other contexts as well so I do think it'd be worthwhile seeing some sort of protection in core as well.

> I would suggest that we support another field in query_vars called strict. If set to true, only then these changes will come into effect, this will allow us to be more liberal with throwing exceptions instead of silently ignoring them and adding to log. wdyt?

Makes sense, although I wonder about the visibility of that for extensions/developers that may find value in it. Practically how likely is it that this would be used? Wouldn't be much harm in implementing it I suppose.

vedanshujain commented:
I am removing the high priority label for now since the immediate case was already handled by the blocks long ago.

barryhughes commented:
Closing, but only to remove it from the PR queue. I'll link to it from #32058 which should help to ensure we tackle this in an upcoming sprint/dev cycle.
